lib/Dialect/Arc/Transforms/LegalizeStateUpdate.cpp

```diff
@@ -126,128 +87,153 @@ struct AccessAnalysis : public DataFlowAnalysis {
+  void add(BlockAccesses &block, Access access);
+};
+} // namespace
 
-  // Aggregate the accesses performed by the operations in this block.
-  SmallPtrSet<Value, 4> localState;
-  AccessState innerAccesses(block);
-  for (Operation &op : *block) {
-    if (isa<AllocStateOp>(&op)) {
-      localState.insert(op.getResult(0));
-      recordState(op.getResult(0));
+LogicalResult AccessAnalysis::analyze(Operation *op) {
+  LLVM_DEBUG(llvm::dbgs() << "Analyzing accesses in " << op->getName() << "\n");
+
+  // Create the lattice nodes for all blocks and operations.
+  llvm::SmallSetVector<OpAccesses *, 16> initWorklist;
+  initWorklist.insert(&get(op));
+  while (!initWorklist.empty()) {
+    OpAccesses &opNode = *initWorklist.pop_back_val();
+
+    // First create lattice nodes for all nested blocks and operations.
+    for (auto &region : opNode.op->getRegions()) {
+      for (auto &block : region) {
+        BlockAccesses &blockNode = get(&block);
+        blockNode.parent = &opNode;
+        for (auto &subOp : block) {
+          if (!isOpInteresting(&subOp))
+            continue;
+          OpAccesses &subOpNode = get(&subOp);
+          if (!subOp.hasTrait<OpTrait::IsIsolatedFromAbove>()) {
+            subOpNode.parent = &blockNode;
+          }
+          initWorklist.insert(&subOpNode);
+        }
+      }
     }
-    if (!isOpInteresting(&op))
-      continue;
-    innerAccesses.join(*getOrCreateFor<AccessState>(block, &op));
-  }
 
-  // Remove any information about locally-defined state which we cannot access
-  // outside the current block. This prevents significant blow-up of the access
-  // sets, since local state accesses don't get transported to parent ops where
-  // they have no meaning.
-  for (auto state : localState) {
-    innerAccesses.remove(state, AccessState::Read);
-    innerAccesses.remove(state, AccessState::Write);
+    // Track the relationship between callers and callees.
+    if (auto callOp = dyn_cast<CallOpInterface>(opNode.op))
+      if (auto calleeOp = callOp.resolveCallable(&symbolTable))
+        get(calleeOp).callers.insert(&opNode);
+
+    // Create the seed accesses.
+    if (auto readOp = dyn_cast<StateReadOp>(opNode.op))
+      add(opNode, Access(readOp.getState(), AccessType::Read));
+    else if (auto writeOp = dyn_cast<StateWriteOp>(opNode.op))
+      add(opNode, Access(writeOp.getState(), AccessType::Write));
   }
+  LLVM_DEBUG(llvm::dbgs() << "- Prepared " << blockAccesses.size()
+                          << " block and " << opAccesses.size()
+                          << " op lattice nodes\n");
+  LLVM_DEBUG(llvm::dbgs() << "- Worklist has " << opWorklist.size()
+                          << " initial ops\n");
+
+  // Propagate accesses through calls.
+  while (!opWorklist.empty()) {
+    auto &opNode = *opWorklist.pop_back_val();
+    if (opNode.callers.empty())
+      continue;
+    auto calleeOp = dyn_cast<CallableOpInterface>(opNode.op);
+    if (!calleeOp)
+      continue;
```

Comment:
Probably! I was thinking that there might be call ops that call some custom op that doesn't impl callable. But we should probably just assert this. Or maybe the call op intf already verifies this.

Reply:
Ah I see, I think it's fair to just error out then. That way we'd also notice immediatly when we've hit some unexpected IR rather than failing silently.